packages/common/src/state/simple.ts

```diff
@@ -0,0 +1,185 @@
+  }
+
+  add() {
+    const newTopA = new Map(this.topA())
+    for (const [address, account] of newTopA) {
+      const accountCopy =
+        account !== undefined
+          ? Object.assign(Object.create(Object.getPrototypeOf(account)), account)
+          : undefined
+      newTopA.set(address, accountCopy)
+    }
+    this.accountStack.push(newTopA)
+    this.codeStack.push(new Map(this.topC()))
+    this.storageStack.push(new Map(this.topS()))
+  }
+
+  async getAccount(address: Address): Promise<Account | undefined> {
+    return this.topA().get(address.toString())
+  }
+
+  async putAccount(address: Address, account?: Account | undefined): Promise<void> {
+    this.topA().set(address.toString(), account)
+  }
+
+  async deleteAccount(address: Address): Promise<void> {
+    this.topA().set(address.toString(), undefined)
+  }
+
+  async modifyAccountFields(address: Address, accountFields: AccountFields): Promise<void> {
+    let account = await this.getAccount(address)
+    if (!account) {
+      account = new Account()
+    }
+    account.nonce = accountFields.nonce ?? account.nonce
+    account.balance = accountFields.balance ?? account.balance
+    account.storageRoot = accountFields.storageRoot ?? account.storageRoot
+    account.codeHash = accountFields.codeHash ?? account.codeHash
+    await this.putAccount(address, account)
+  }
+
+  async getContractCode(address: Address): Promise<Uint8Array> {
+    return this.topC().get(address.toString()) ?? new Uint8Array(0)
+  }
+
+  async putContractCode(address: Address, value: Uint8Array): Promise<void> {
+    this.topC().set(address.toString(), value)
+    if ((await this.getAccount(address)) === undefined) {
+      await this.putAccount(address, new Account())
+    }
+    await this.modifyAccountFields(address, { codeHash: keccak256(value) })
+  }
+
+  async getContractCodeSize(address: Address): Promise<number> {
+    const contractCode = await this.getContractCode(address)
+    return contractCode.length
+  }
+
+  async getContractStorage(address: Address, key: Uint8Array): Promise<Uint8Array> {
+    return this.topS().get(`${address.toString()}_${bytesToHex(key)}`) ?? new Uint8Array(0)
+  }
+
+  async putContractStorage(address: Address, key: Uint8Array, value: Uint8Array): Promise<void> {
+    this.topS().set(`${address.toString()}_${bytesToHex(key)}`, value)
+  }
+
+  async checkpoint(): Promise<void> {
+    this.add()
+  }
+  async commit(): Promise<void> {
+    const topA = this.accountStack.pop()
+    const topC = this.codeStack.pop()
+    const topS = this.storageStack.pop()
+    this.accountStack.pop()
+    this.codeStack.pop()
+    this.storageStack.pop()
+    this.accountStack.push(topA!)
+    this.codeStack.push(topC!)
+    this.storageStack.push(topS!)
+  }
+
+  async revert(): Promise<void> {
+    this.accountStack.pop()
+    this.codeStack.pop()
+    this.storageStack.pop()
+  }
+
+  async flush(): Promise<void> {}
+  clearCaches(): void {}
```

Comment:
These empty methods should throw a "Not implemented" error.

Reply:
Ah, yeah, as you seemed to have noted somewhere below, these methods are called from within the EVM so they need to be callable.

General note: One thing that this state manager work further exposes is that our whole interface design around the state manger is not optimal, so what is generally in and what not, what is optional, around what principles the different interfaces are structured. This needs some refactoring along the breaking releases, likely together with the implementations itself. Will write up some issue with some evolved thinking about this at some point.

Reply:
Ah ok, good point that these need to be callable, and also good point regarding the interface.

Reply:
Yeah, there are various small optimizations/changes we can make to the interface which in sum will make them significantly more solid, so in these two cases e.g. it makes a lot of sense to make both methods optional in the interfaces and only call with TypeScript's '?'.

Small thing admittedly, but these kind of single things add up.